NeoOnnx/src/LayerNormFusionOptimizer.cpp

```diff
@@ -0,0 +1,125 @@
+/* Copyright © 2017-2020 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <cstring>
+#include "LayerNormFusionOptimizer.h"
+
+namespace NeoOnnx {
+
+void CLayerNormFusionOptimizer::Apply()
+{
+	//graph.Reshape(); // be sure, the all BLOBs' sizes are inited in the graph
+
+	int normLayersCount = 0;
+	CArray<const char*> layersList{};
+	graph.GetLayerList( layersList );
```

Comment:
Here' s the problem. The `layerList` isn't guaranteed to be in topological sorted order. E.g. if you make `GetLayerList` after your transformation then newly added layers will be _at the end_ of the array. As a result, when you detecting block you may delete layer before iterating over it (imagine, `bias` is the last layer in the list).

The fastest wawy to fix it (as I see) is to check `HasLayer` (otherwise you will get an exception in `GetLayer`).

Reply:
There is no problem. I just iterating all the nodes in graph. Inside the loop I'm walking by inputLayer link in the CBaseLayer class. I do not consider that the layersList is topologically sorted.

Reply:
No, the problem is the following: there is a chance that in `layersList` the block will be located in order the following order `{ ..., addLayerName, biasLayerName, scaleLayerName, ... }`. In that case when `layerList[i] == addLayerName` you will remove the whole block from the graph. But not from the `layerList`.

Then, on the next iteration you `layerList[i+1] == biasLayerName` which has already been removed from the `Graph`. As a result `Graph.GetLayer( layerList[i] )` will throw an exception.

Reply:
The situation is even worse. `layerList` consists of `const char*` which don't own the data. If the layer is deleted, it could be destroyed and `layerList[i+1]` may lead to freed memory. Even `HasLayer` doesn't look like a good option...

Reply:
Yes, you are right. This is possible. My suggestion is to null those items in `layerList`, which are already removed from the graph, and to skip loop iterations on them.